jalib/jfilesystem.cpp

```diff
@@ -151,6 +151,46 @@ jalib::Filesystem::DirName(const dmtcp::string &str)
   return str.substr(0, lastSlash);
 }
 
+void jalib::Filesystem::DirName(char *dirname, const char *fname)
+{
+  size_t len = strlen(fname);
+
+  if (strcmp(fname, "/") == 0 || strcmp(fname, ".") == 0 || fname[0] == '\0') {
+    strcpy(dirname, fname);
+    return;
+  }
+
+  if (strcmp(fname, "..") == 0) {
+    strcpy(dirname, ".");
+    return;
+  }
+
+  // Remove trailing slashes
+  while (len > 0 && fname[len - 1] == '/') {
+    len--;
+  }
+
+  size_t lastSlash = 0;
```

Comment:
I suggest: `int lastSlash = -1;`  It means that you will also change to:
> `for (int i = 0; i < len; i++) {`
below.  The constant 0 could be confused by the reader with index 0, even though you test for `fname == '/' `above.